File: android/src/main/java/expo/modules/arcgismapssdk/CustomDynamicEntityFeed.kt
package expo.modules.arcgismapssdk

import com.arcgismaps.data.Field
import com.arcgismaps.data.FieldType
import com.arcgismaps.geometry.Point
import com.arcgismaps.geometry.SpatialReference
import com.arcgismaps.mapping.layers.DynamicEntityLayer
import com.arcgismaps.realtime.CustomDynamicEntityDataSource
import com.arcgismaps.realtime.DynamicEntityDataSourceInfo
import expo.modules.arcgismapssdk.dto.CustomDynamicEntityFeedRecord
import java.io.File
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import org.json.JSONObject

/**
 * A [CustomDynamicEntityDataSource.EntityFeedProvider] that replays observations
 * from a local newline-delimited JSON (JSONL) file: each line is a flat object
 * of attributes that must include the entity-id, longitude, and latitude fields.
 * Field types are inferred from the first observation. Observations are emitted
 * on [onConnect] at [observationsPerSecond], looping so the layer keeps moving.
 */
class LocalFileEntityFeedProvider(
  private val record: CustomDynamicEntityFeedRecord,
  private val scope: CoroutineScope,
) : CustomDynamicEntityDataSource.EntityFeedProvider {
  private val observations: List<JSONObject> = parseJsonl(record.observationsPath)
  private val feedFlow =
    MutableSharedFlow<CustomDynamicEntityDataSource.FeedEvent>(extraBufferCapacity = 64)

  override val feed: SharedFlow<CustomDynamicEntityDataSource.FeedEvent>
    get() = feedFlow

  override suspend fun onLoad(): DynamicEntityDataSourceInfo {
    val first = observations.firstOrNull() ?: JSONObject()
    return DynamicEntityDataSourceInfo(record.entityIdField, inferFields(first)).apply {
      spatialReference = SpatialReference.wgs84()
    }
  }

  override suspend fun onConnect() {
    if (observations.isEmpty()) return
    val perSecond = record.observationsPerSecond ?: 10.0
    val periodMs = if (perSecond > 0) (1000.0 / perSecond).toLong().coerceAtLeast(1) else 100
    scope.launch(Dispatchers.Default) {
      var index = 0
      while (isActive) {
        val attributes = observations[index % observations.size]
        emitObservation(attributes)
        index++
        delay(periodMs)
      }
    }
  }

  override suspend fun onDisconnect() {}

  private suspend fun emitObservation(attributes: JSONObject) {
    val lon = attributes.optDouble(record.longitudeField, Double.NaN)
    val lat = attributes.optDouble(record.latitudeField, Double.NaN)
    if (lon.isNaN() || lat.isNaN()) return
    val point = Point(lon, lat, SpatialReference.wgs84())
    val map = mutableMapOf<String, Any>()
    for (key in attributes.keys()) {
      if (key == record.longitudeField || key == record.latitudeField) continue
      map[key] = attributes.get(key)
    }
    feedFlow.emit(CustomDynamicEntityDataSource.FeedEvent.NewObservation(point, map))
  }

  /** Numeric attributes become float64 fields; everything else text. */
  private fun inferFields(observation: JSONObject): List<Field> =
    observation.keys().asSequence().map { key ->
      val type = if (observation.get(key) is Number) FieldType.Float64 else FieldType.Text
      Field(type, key, key, 0)
    }.toList()

  private fun parseJsonl(path: String): List<JSONObject> {
    val file = File(path)
    if (!file.exists()) return emptyList()
    return file
      .readLines()
      .mapNotNull { line ->
        val trimmed = line.trim()
        if (trimmed.isEmpty()) null else runCatching { JSONObject(trimmed) }.getOrNull()
      }
  }
}

/**
 * Builds a [DynamicEntityLayer] backed by a [CustomDynamicEntityDataSource] that
 * replays a local JSONL file, and connects it. Returns null if the file is empty
 * or unreadable.
 */
fun makeCustomDynamicEntityLayer(
  feed: CustomDynamicEntityFeedRecord,
  scope: CoroutineScope,
): DynamicEntityLayer? {
  if (!File(feed.observationsPath).exists()) return null
  val provider = LocalFileEntityFeedProvider(feed, scope)
  val source = CustomDynamicEntityDataSource(provider)
  return DynamicEntityLayer(source).also { scope.launch { source.connect() } }
}

File: ios/CustomDynamicEntityFeed.swift
import ArcGIS
import Foundation

/// A {@link CustomDynamicEntityFeed} that replays observations from a local
/// newline-delimited JSON (JSONL) file. Each line is a flat JSON object of
/// attributes that must include the entity-id, longitude, and latitude fields.
/// Field types are inferred from the first observation.
struct LocalFileDynamicEntityFeed: CustomDynamicEntityFeed {
  let observations: [[String: Any]]
  let longitudeField: String
  let latitudeField: String
  let interval: TimeInterval

  /// Emits each observation as a `newObservation` event, spaced by `interval`,
  /// looping back to the start so the layer keeps animating.
  var events: AsyncStream<CustomDynamicEntityFeedEvent> {
    AsyncStream { continuation in
      let task = Task {
        guard !observations.isEmpty else {
          continuation.finish()
          return
        }
        var index = 0
        while !Task.isCancelled {
          let attributes = observations[index % observations.count]
          if let point = makePoint(attributes) {
            continuation.yield(.newObservation(geometry: point, attributes: sendableAttributes(attributes)))
          }
          index += 1
          try? await Task.sleep(nanoseconds: UInt64(interval * 1_000_000_000))
        }
        continuation.finish()
      }
      continuation.onTermination = { _ in task.cancel() }
    }
  }

  /// Converts a JSON attribute dictionary to concrete `Sendable` values (numbers
  /// to Double, strings to String), dropping the coordinate fields (they live in
  /// the geometry) and any values of other types.
  private func sendableAttributes(_ attributes: [String: Any]) -> [String: any Sendable] {
    var result: [String: any Sendable] = [:]
    for (key, value) in attributes where key != longitudeField && key != latitudeField {
      if let string = value as? String {
        result[key] = string
      } else if let number = value as? NSNumber {
        result[key] = number.doubleValue
      }
    }
    return result
  }

  private func makePoint(_ attributes: [String: Any]) -> Point? {
    guard let lon = numeric(attributes[longitudeField]),
      let lat = numeric(attributes[latitudeField])
    else { return nil }
    return Point(x: lon, y: lat, spatialReference: .wgs84)
  }
}

/// Coerce a JSON value to a Double (numbers arrive as NSNumber; numeric strings
/// are also accepted).
private func numeric(_ value: Any?) -> Double? {
  if let number = value as? NSNumber { return number.doubleValue }
  if let string = value as? String { return Double(string) }
  return nil
}

/// Builds a `DynamicEntityLayer` backed by a `CustomDynamicEntityDataSource`
/// that replays a local JSONL file. Returns nil if the file can't be read or is
/// empty. The data source's field schema is inferred from the first observation.
func makeCustomDynamicEntityLayer(_ feed: CustomDynamicEntityFeedRecord) -> DynamicEntityLayer? {
  guard let data = FileManager.default.contents(atPath: feed.observationsPath) else { return nil }
  let observations = parseJSONL(data)
  guard let first = observations.first else { return nil }

  let fields = inferFields(from: first)
  let info = DynamicEntityDataSourceInfo(entityIDFieldName: feed.entityIdField, fields: fields)
  info.spatialReference = .wgs84

  let perSecond = feed.observationsPerSecond ?? 10
  let interval = perSecond > 0 ? 1 / perSecond : 0.1

  let source = CustomDynamicEntityDataSource(info: info) {
    LocalFileDynamicEntityFeed(
      observations: observations,
      longitudeField: feed.longitudeField,
      latitudeField: feed.latitudeField,
      interval: interval)
  }
  let layer = DynamicEntityLayer(dataSource: source)
  Task { [weak source] in try? await source?.connect() }
  return layer
}

/// Parses newline-delimited JSON into an array of flat attribute dictionaries.
private func parseJSONL(_ data: Data) -> [[String: Any]] {
  guard let text = String(data: data, encoding: .utf8) else { return [] }
  var result: [[String: Any]] = []
  for line in text.split(whereSeparator: \.isNewline) {
    let trimmed = line.trimmingCharacters(in: .whitespaces)
    guard !trimmed.isEmpty, let lineData = trimmed.data(using: .utf8),
      let object = try? JSONSerialization.jsonObject(with: lineData) as? [String: Any]
    else { continue }
    result.append(object)
  }
  return result
}

/// Infers an ArcGIS field schema from the first observation: numeric values
/// become `float64` fields, everything else `text`.
private func inferFields(from observation: [String: Any]) -> [Field] {
  observation.map { key, value in
    let type: FieldType = value is NSNumber ? .float64 : .text
    return Field(type: type, name: key, alias: key, length: 0)
  }
}
